Add xla_cpu_ftz flag to control flush-to-zero on CPU

yashwantbezawada opened the pull request:
Fixes #22858

I ran into needing to control denormal handling on CPU similar to how
xla_gpu_ftz works for GPU. Currently the CPU backend unconditionally sets
the denormal-fp-math=preserve-sign attribute on all functions.

This adds an xla_cpu_ftz flag (field 493 in DebugOptions) that controls
whether this attribute gets applied in CreateCpuFunction. It defaults to
true so existing behavior is preserved. Users who want to disable
flush-to-zero can set --xla_cpu_ftz=false.

Usage:
```
XLA_FLAGS="--xla_cpu_ftz=false" ./program
```

yashwantbezawada committed:
Add xla_cpu_ftz flag to control flush-to-zero on CPU

Fixes #22858

Needed a way to control denormal handling on CPU similar to xla_gpu_ftz for GPU. The CPU backend was unconditionally setting denormal-fp-math=preserve-sign on all functions.

This flag controls whether that attribute gets applied. Defaults to true to keep existing behavior, set --xla_cpu_ftz=false to disable.

Also added unit tests as requested.

penpornk commented on xla/service/llvm_ir/llvm_util.cc:
```diff
@@ -768,10 +768,11 @@ llvm::Function* CreateCpuFunction(llvm::FunctionType* function_type,
+  // NEON instruction set. Controlled by the xla_cpu_ftz flag.
+  if (module_config.debug_options().xla_cpu_ftz()) {
+    function->addFnAttr("denormal-fp-math", "preserve-sign");
+  }
```
Would you mind adding a unit test for this? E.g., set the flag to true/false and check for attributes.
https://github.com/openxla/xla/blob/189a63c1e028d7d0caf7dde17b7323ce480aabf7/xla/service/gpu/tests/gpu_codegen_test.cc#L42

penpornk reviewed (changes requested):
Thank you for the PR!

yashwantbezawada commented:
Rebased onto latest main and resolved the merge conflict in the test file. The unit tests for the flag are still in place. Ready for another look when you get a chance.

dimitar-asenov commented:
@penpornk Could you please review this PR?

penpornk reviewed (commented):
Thank you for adding the test and I'm so sorry for the delay!

@seantalts is looking at CPU numerics and the flags to control them and is the better reviewer for this PR.

seantalts reviewed (approved):
lgtm but could you please rebase? ty!

yashwantbezawada commented:
@seantalts Rebased onto latest main. Field 443 got reserved in DebugOptions since I last updated this, so I moved xla_cpu_ftz to 493 and bumped the next id comment. The test file also moved to restricted/ on main, so the tests went along with it. No other changes, should be good to go now.
